File: tests/providers/context.test.ts
import { describe, it, expect } from '@jest/globals';
import { contextProvider } from '../../src/providers/context.js';
import type { SessionData } from '../../src/types.js';

describe('context provider', () => {
  it('formats tokens as K for thousands', async () => {
    const session: SessionData = {
      cwd: '/tmp',
      context_window: {
        used_percentage: 42,
        current_usage: { input_tokens: 38000, cache_creation_input_tokens: 2000, cache_read_input_tokens: 1500 },
      },
    };
    const data = await contextProvider.resolve(session) as any;
    expect(data.tokens).toBe('42K');
    expect(data.percent).toBe(42);
  });

  it('formats tokens as M for millions', async () => {
    const session: SessionData = {
      cwd: '/tmp',
      context_window: {
        used_percentage: 85,
        current_usage: { input_tokens: 1000000, cache_creation_input_tokens: 100000, cache_read_input_tokens: 100000 },
      },
    };
    const data = await contextProvider.resolve(session) as any;
    expect(data.tokens).toBe('1.2M');
  });

  it('returns raw number for small token counts', async () => {
    const session: SessionData = {
      cwd: '/tmp',
      context_window: {
        used_percentage: 1,
        current_usage: { input_tokens: 500, cache_creation_input_tokens: 0, cache_read_input_tokens: 0 },
      },
    };
    const data = await contextProvider.resolve(session) as any;
    expect(data.tokens).toBe('500');
  });

  it('returns null fields when context_window missing', async () => {
    const session: SessionData = { cwd: '/tmp' };
    const data = await contextProvider.resolve(session) as any;
    expect(data.tokens).toBeNull();
    expect(data.percent).toBeNull();
  });

  it('has correct name', () => {
    expect(contextProvider.name).toBe('context');
  });
});

File: internal/provider/context_test.go
package provider

import (
	"testing"

	"github.com/jheddings/ccglow/internal/types"
)

func TestFormatTokens(t *testing.T) {
	tests := []struct {
		input    int
		expected string
	}{
		{0, "0"},
		{500, "500"},
		{1000, "1K"},
		{24500, "24K"},
		{1000000, "1M"},
		{1500000, "1.5M"},
		{2000000, "2M"},
	}

	for _, tt := range tests {
		result := FormatTokens(tt.input)
		if result != tt.expected {
			t.Errorf("FormatTokens(%d) = %q, want %q", tt.input, result, tt.expected)
		}
	}
}

func contextValues(result *types.ProviderResult) map[string]any {
	return result.Values["context"].(map[string]any)
}

func contextPercent(result *types.ProviderResult) map[string]any {
	return contextValues(result)["percent"].(map[string]any)
}

func TestContextProvider(t *testing.T) {
	p := &contextProvider{}
	sess := &types.SessionData{
		CWD: "/tmp",
		ContextWindow: &types.ContextWindow{
			UsedPercentage:    36,
			ContextWindowSize: 1000000,
			CurrentUsage: &types.CurrentUsage{
				InputTokens:              100,
				CacheCreationInputTokens: 200,
				CacheReadInputTokens:     300,
			},
		},
	}

	result, err := p.Resolve(sess)
	if err != nil {
		t.Fatal(err)
	}

	ctx := contextValues(result)
	if ctx["tokens"] != "600" {
		t.Errorf("expected 600 tokens, got %s", ctx["tokens"])
	}
	if ctx["size"] != "1M" {
		t.Errorf("expected 1M size, got %s", ctx["size"])
	}
	pct := contextPercent(result)
	if pct["used"] != 36 {
		t.Errorf("expected 36%%, got %v", pct["used"])
	}
}

func TestContextProviderWithTotalTokens(t *testing.T) {
	p := &contextProvider{}
	inputTokens := 50000
	outputTokens := 8000
	sess := &types.SessionData{
		CWD: "/tmp",
		ContextWindow: &types.ContextWindow{
			UsedPercentage:    36,
			ContextWindowSize: 1000000,
			TotalInputTokens:  &inputTokens,
			TotalOutputTokens: &outputTokens,
			CurrentUsage: &types.CurrentUsage{
				InputTokens:              100,
				CacheCreationInputTokens: 200,
				CacheReadInputTokens:     300,
			},
		},
	}

	result, err := p.Resolve(sess)
	if err != nil {
		t.Fatal(err)
	}

	ctx := contextValues(result)
	if ctx["input"] != "50K" {
		t.Errorf("expected Input 50K, got %s", ctx["input"])
	}
	if ctx["output"] != "8K" {
		t.Errorf("expected Output 8K, got %s", ctx["output"])
	}
}

func TestContextProviderInputFallback(t *testing.T) {
	p := &contextProvider{}
	sess := &types.SessionData{
		CWD: "/tmp",
		ContextWindow: &types.ContextWindow{
			UsedPercentage: 10,
			CurrentUsage: &types.CurrentUsage{
				InputTokens:              100,
				CacheCreationInputTokens: 200,
				CacheReadInputTokens:     300,
			},
		},
	}

	result, err := p.Resolve(sess)
	if err != nil {
		t.Fatal(err)
	}

	ctx := contextValues(result)
	if ctx["input"] != "600" {
		t.Errorf("expected Input 600, got %s", ctx["input"])
	}
	if ctx["output"] != "" {
		t.Errorf("expected empty Output, got %s", ctx["output"])
	}
}

func TestContextProviderRemaining(t *testing.T) {
	p := &contextProvider{}
	sess := &types.SessionData{
		CWD: "/tmp",
		ContextWindow: &types.ContextWindow{
			UsedPercentage:      36,
			RemainingPercentage: 64,
			ContextWindowSize:   1000000,
			CurrentUsage:        &types.CurrentUsage{InputTokens: 100},
		},
	}

	result, err := p.Resolve(sess)
	if err != nil {
		t.Fatal(err)
	}

	pct := contextPercent(result)
	if pct["remaining"] != 64 {
		t.Errorf("expected remaining 64, got %v", pct["remaining"])
	}
}

func TestContextProviderNoRemaining(t *testing.T) {
	p := &contextProvider{}
	sess := &types.SessionData{CWD: "/tmp"}

	result, err := p.Resolve(sess)
	if err != nil {
		t.Fatal(err)
	}

	pct := contextPercent(result)
	if pct["remaining"] != 0 {
		t.Errorf("expected remaining 0, got %v", pct["remaining"])
	}
}

func TestContextProviderZeroRemaining(t *testing.T) {
	p := &contextProvider{}
	sess := &types.SessionData{
		CWD:           "/tmp",
		ContextWindow: &types.ContextWindow{},
	}

	result, err := p.Resolve(sess)
	if err != nil {
		t.Fatal(err)
	}

	pct := contextPercent(result)
	if pct["remaining"] != 0 {
		t.Errorf("expected remaining 0 for zero value with no usage, got %v", pct["remaining"])
	}
}
